File: C/skiplist/sset.c
/**
 * Author: Axel Ariel Saravia
 */

/*
    IMPORTANT

    the T type is the signed type

    skiplinkSSet struct is {
        size_t height;
        skiplinkSSet** next;
        T value;
    }

    skiplistSSet struct is {
        size_t height;
        size_t length;
        skiplinkSSet* sentinel;
        skiplinkSSet* stack[32];
    }
*/

# include "sset.h"

skiplinkSSet* skipLink_create(T const value, size_t const height) {
    skiplinkSSet* n = malloc(sizeof(skiplinkSSet));
    skiplinkSSet** next =  malloc(sizeof(skiplinkSSet*[height + 1]));
    if (!n || !next) {
        return 0;
    }
    *n = (skiplinkSSet){
        .height = height,
        .next = next,
        .value = value,
    };
    return n;
}

skiplistSSet* skiplistSSet_init(skiplistSSet list[static 1]) {
    skiplinkSSet* sentinel = skipLink_create(0, MAX_HEIGHT);
    if (!sentinel) {
        return 0;
    }
    *list = (skiplistSSet){
        .sentinel = sentinel,
    };
    return list;
}

skiplinkSSet* skiplistSSet_get_pred_node(skiplistSSet const list[static 1], T const value) {
    skiplinkSSet* node = list->sentinel;
    size_t r = list->height;
    while (r >= 0) {
        while (node->next[r] && node->next[r]->value < value) {
            node = node->next[r];
        }
        r -= 1;
    }
    return node;
}

size_t skiplistSSet_size(skiplistSSet list[static 1]) {
    return list->length;
}

T skiplistSSet_find(skiplistSSet const* list, T const value) {
    skiplinkSSet* node = skiplistSSet_get_pred_node(list, value);
    return (
        !node || !node->next[0]
        ? 0
        : node->next[0]->value
    );
}

bool skiplistSSet_remove(skiplistSSet list[static 1], T const value) {
    bool removed = false;
    skiplinkSSet* node = list->sentinel;
    skiplinkSSet* temp = {0};
    signed r = list->height;
    while (r >= 0) {
        while (node->next[r] && node->next[r]->value < value) {
            node = node->next[r];
        }
        if (node->next[r] && node->next[r]->value == value) {
            temp = node->next[r];
            removed = true;
            node->next[r] = node->next[r]->next[r];
            if (node == list->sentinel && !node->next[r]) {
                list->height -= 1;
            }
        }
        r -= 1;
    }
    if (removed) {
        list->length -= 1;
        free(temp->next);
        free(temp);
    }
    return removed;
}

size_t skiplistSSet_pick_height(void) {
    srand(time(0));
    size_t const z = rand();
    size_t k = 0;
    size_t m = 1;
    while (z & m) {
        k += 1;
        m <<= 1;
    }
    return k;
}

bool skiplistSSet_add(skiplistSSet list[static 1], T const value) {
    skiplinkSSet* node = list->sentinel;
    signed r = list->height;
    while (r >= 0) {
        while (node->next[r] && node->next[r]->value < value) {
            node = node->next[r];
        }
        if (node->next[r] && node->next[r]->value == value) {
           return false;
        }
        list->stack[r] = node;
        r -= 1;
    }

    size_t const height = skiplistSSet_pick_height();
    skiplinkSSet* new_node = skipLink_create(value, height);
    if (!new_node) {
        return false;
    }

    while (list->height < height) {
        list->height += 1;
        list->stack[list->height] = list->sentinel;
    }
    for (size_t i = 0; i <= height; i += 1) {
        new_node->next[i] = list->stack[i]->next[i];
        list->stack[i]->next[i] = new_node;
    }
    list->length += 1;
    return true;
}

void skiplistSSet_reset(skiplistSSet list[static 1]) {
    skiplinkSSet* node = list->sentinel->next[0];
    while (node) {
        skiplinkSSet* temp = node;
        node = node->next[0];
        free(temp->next);
        free(temp);
    }
    signed r = list->height;
    while (r >= 0) {
        list->sentinel->next[r] = 0;
        r -= 1;
    }
    list->height = 0;
    list->length = 0;
}

void skiplistSSet_free(skiplistSSet list[static 1]) {
    skiplinkSSet* node = list->sentinel;
    while (node) {
        skiplinkSSet* temp = node;
        node = node->next[0];
        free(temp->next);
        free(temp);
    }
    *list = (skiplistSSet){0};
}

File: Go/skipList/sset.go
/**
* Author: Axel Ariel Saravia
*/

package skiplist

type SSetNode[T inequable] struct {
    Height uint
    Next []*SSetNode[T]
    Value T
}

type SSet[T inequable] struct {
    Height uint
    Len uint
    Sentinel SSetNode[T]
    Stack []*SSetNode[T]
}

func NewSSetNode[T inequable](v T, h uint) *SSetNode[T] {
    var res *SSetNode[T] = new(SSetNode[T])
    *res = SSetNode[T]{
        Height: h,
        Next: make([]*SSetNode[T], h + 1),
        Value: v,
    }
    return res
}

// MAX_HEIGHT = 31
func CreateSSet[T inequable]() SSet[T] {
    return SSet[T]{
        Height: 0,
        Len: 0,
        Sentinel: SSetNode[T]{
            Height: MAX_HEIGHT,
            Next: make([]*SSetNode[T], MAX_HEIGHT + 1),
        },
        Stack: make([]*SSetNode[T], MAX_HEIGHT),
    }
}

func (s *SSet[T]) GetNode(v T) *SSetNode[T] {
    sn := &s.Sentinel
    r := int(s.Height)
    for r >= 0 {
        for sn.Next[r] != nil && sn.Next[r].Value < v {
            sn = sn.Next[r]
        }
        r -= 1
    }
    return sn.Next[0]
}

func (s *SSet[T]) Find(v T) (T, *SSetNode[T]) {
    sn := s.GetNode(v)
    var res T 
    if sn != nil {
        res = sn.Value
    }
    return res, sn
}

func (s *SSet[T]) Add(v T) bool {
    sn := &s.Sentinel
    r := int(s.Height)
    for r >= 0 {
        for sn.Next[r] != nil && sn.Next[r].Value < v {
            sn = sn.Next[r]
        }
        if sn.Next[r] != nil && sn.Next[r].Value == v {
            return false
        }
        s.Stack[r] = sn
        r -= 1
    }
    newh := pickHeight()
    newsn := NewSSetNode[T](v, newh)
    for s.Height < newh {
        s.Height += 1
        s.Stack[s.Height] = &s.Sentinel
    }
    var i uint = 0
    for i < newh {
        newsn.Next[i] = s.Stack[i].Next[i]
        s.Stack[i].Next[i] = newsn
        i += 1
    }
    s.Len += 1
    return true
}

func (s *SSet[T]) Remove(v T) bool {
    removed := false
    sn := &s.Sentinel
    r := s.Height
    for r >= 0 {
        for sn.Next[r] != nil && sn.Next[r].Value < v {
            sn = sn.Next[r]
        }
        if sn.Next[r] != nil && sn.Next[r].Value == v {
            removed = true
            sn.Next[r] = sn.Next[r].Next[r]
            if sn == &s.Sentinel && sn.Next[r] == nil {
                s.Height -= 1
            }
        }
        r -= 1
    }
    if removed {
        s.Len -= 1
    }
    return removed
}
